### PR TITLE
GH #14499 Panel.ffill ignores axis parameter and fill along axis=1

### LINKED ISSUE #8251: Panel.fillna with method='ffill' ignores the axis parameter and only fills along axis=1
This is also a documentation error, as the docs say the default axis is axis = 0

```
import pandas as pd
from pandas.util.testing import makePanel

#
# Demo A: We want to fill along axis 0
# 
p = makePanel().transpose(1,0,2)
p.iloc[10:20,0,0] = np.nan

# Demonstrate we have a gap
p.iloc[:,0,0].plot()

# Demonstrate that nothing seems to fill the gap
p.fillna(method='ffill').iloc[:,0,0].plot()
p.fillna(method='ffill', axis=0).iloc[:,0,0].plot()
p.fillna(method='ffill', axis=1).iloc[:,0,0].plot()
p.fillna(method='ffill', axis=2).iloc[:,0,0].plot()

# At least one of the above should be the same as:
p.apply(lambda series: series.fillna(method='ffill'), axis=0).iloc[:,0,0].plot()

#
# Demo B: We want to fill along axis 1
#
p = makePanel()
p.iloc[0, 10:20,0] = np.nan

# Demonstrate we have a gap
p.iloc[0,:,0].plot()

# Demonstrate the gap is filled regardless of the axis
p.fillna(method='ffill').iloc[0,:,0].plot()
p.fillna(method='ffill', axis=0).iloc[0,:,0].plot()
p.fillna(method='ffill', axis=1).iloc[0,:,0].plot()
p.fillna(method='ffill', axis=2).iloc[0,:,0].plot()
```

### LINKED ISSUE #14499: Panel.ffill ignores axis parameter and fill along axis=1
#### A small, complete example of the issue

``` python
a = pd.Panel({'a':np.arange(15).reshape((5,3)), 'b':np.arange(15,30).reshape((5,3))})
a['a'].loc[1,1] = np.nan

a['a']
Out[36]:
    0     1   2
0   0   1.0   2
1   3   NaN   5
2   6   7.0   8
3   9  10.0  11
4  12  13.0  14

a.ffill(axis=2)['a']
Out[38]:
      0     1     2
0   0.0   1.0   2.0
1   3.0   1.0   5.0
2   6.0   7.0   8.0
3   9.0  10.0  11.0
4  12.0  13.0  14.0
```
#### Expected Output

`a.at['a', 1, 1] == 3`
If you ffill along axis=2, you should get `3` instead of `1`
#### Output of `pd.show_versions()`

Pandas 0.19.0

### PR DESCRIPTION
**Pull request by Tux1**
- [x] closes #14499, closes #8251
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -3233,14 +3233,20 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
-                result = dict([(col, s.fillna(method=method, value=value))
+                result = dict([(col, s.fillna(method=method, 
+                                              value=value, 
+                                              axis=axis-1))
                                for col, s in self.iteritems()])
                 new_obj = self._constructor.\
                     from_dict(result).__finalize__(self)
                 new_data = new_obj._data
 
+                    def test_ffill_bfill_axis(self):
```
Comment (jreback) [inline pandas/core/generic.py]: ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (Tux1) [inline pandas/core/generic.py]: fixed

**Comment by Tux1** [comment]
handle case where `axis=0`, tests added for every case and passed, whatsnew added

**Comment by Tux1** [comment]
@jreback you're okay with this pr ?

File: doc/source/whatsnew/v0.19.1.txt [hunk 2]
```diff
@@ -83,4 +83,6 @@ Bug Fixes
 
 
 - Bug in ``pd.pivot_table`` may raise ``TypeError`` or ``ValueError`` when ``index`` or ``columns``
-  is not scalar and ``values`` is not specified (:issue:`14380`)
\ No newline at end of file
+  is not scalar and ``values`` is not specified (:issue:`14380`)
+
+- Bug in ``pd.Panel.ffill`` where ``axis`` was not propagated (:issue:`14499`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.1.txt]: Can you move this to the 0.20.0 file? (0.19.1 is released in the meantime)

File: pandas/tests/test_panel.py [hunk 3]
```diff
@@ -1475,7 +1475,7 @@ def test_fillna(self):
 
         panel = self.panel.copy()
         panel['str'] = 'foo'
-
+        print(panel)
```
Comment (jorisvandenbossche) [inline pandas/tests/test_panel.py]: left over

File: pandas/core/generic.py [hunk 4]
```diff
@@ -3233,21 +3228,41 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
 
             # 3d
             elif self.ndim == 3:
-
                 # fill in 2d chunks
-                result = dict([(col, s.fillna(method=method, value=value))
-                               for col, s in self.iteritems()])
+                if axis == 0:
+                    frame = self.swapaxes(0, 1)
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: Question (not too familiar with Panels): swapping the axes here can potentially cause dtypes to change? (ints upcast to float, or numeric upcast to object when you have mixed dtypes)

Comment (Tux1) [inline pandas/core/generic.py]: Good point. Let me check

File: doc/source/whatsnew/v0.19.1.txt (hunk 2 above)
Comment (Tux1) [inline doc/source/whatsnew/v0.19.1.txt]: done.

**Comment by jreback** [comment]
@Tux1 can you rebase / update

**Comment by Tux1** [comment]
Done

File: pandas/core/generic.py [hunk 5]
```diff
@@ -3262,26 +3262,21 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
                             'you passed a "{0}"'.format(type(value).__name__))
         self._consolidate_inplace()
 
-        # set the default here, so functions examining the signaure
+        # set the default here, so functions examining the signature
         # can detect if something was set (e.g. in groupby) (GH9221)
         if axis is None:
-            axis = 0
```
Comment (jreback) [inline pandas/core/generic.py]: ``axis = self._stat_axis_number`` give the correct default

File: pandas/core/generic.py [hunk 6]
```diff
@@ -3290,21 +3285,41 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
-
                 # fill in 2d chunks
-                result = dict([(col, s.fillna(method=method, value=value))
-                               for col, s in self.iteritems()])
+                if axis == 0:
+                    frame = self.swapaxes(0, 1)
+                    axis2d = 0
+                else:
+                    frame = self
+                    axis2d = axis - 1
```
Comment (jreback) [inline pandas/core/generic.py]: ``axis == 2``?

File: pandas/core/generic.py [hunk 7]
```diff
@@ -3290,21 +3285,41 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
             else:
                 # 2d or less
-                method = missing.clean_fill_method(method)
-                new_data = self._data.interpolate(method=method, axis=axis,
-                                                  limit=limit, inplace=inplace,
-                                                  coerce=True,
-                                                  downcast=downcast)
+                if self._is_mixed_type and axis == 1:
+                    if inplace:
+                        raise NotImplementedError()
```
Comment (jreback) [inline pandas/core/generic.py]: add a message to the ``NotImplementedError``

File: pandas/tests/test_panel.py [hunk 8]
```diff
@@ -1519,6 +1519,50 @@ def test_ffill_bfill(self):
...
+            'a': np.arange(15, dtype=float).reshape((5, 3)),
+            'b': np.arange(15, 30, dtype=float).reshape((5, 3))
+        })
+        f0 = a.copy()
+        f1 = a.copy()
+        f2 = a.copy()
+        b0 = a.copy()
+        b1 = a.copy()
+        b2 = a.copy()
+        a['a'].loc[1, 1] = np.nan
```
Comment (jreback) [inline pandas/tests/test_panel.py]: set using full ``.loc``, not chained indexing, e.g. ``f0.loc['a', 1, 1] = np.nan``

File: pandas/tests/test_panel.py [hunk 9]
```diff
@@ -1519,6 +1519,50 @@ def test_ffill_bfill(self):
...
+        f0['a'].loc[1, 1] = np.nan
+        f1['a'].loc[1, 1] = 1
+        f2['a'].loc[1, 1] = 3
+        b0['a'].loc[1, 1] = 19.0
+        b1['a'].loc[1, 1] = 7
+        b2['a'].loc[1, 1] = 5
+
+        # method='ffill'
+        # axis=0
+        assert_panel_equal(a.ffill(axis=0), f0)
```
Comment (jreback) [inline pandas/tests/test_panel.py]: make these expected before each statement, e.g.

```
result = a.ffill(axis=0)
expected = ......
assert_panel_equal(result, expected)
```

File: pandas/tests/test_panel.py [hunk 10]
```diff
@@ -1519,6 +1519,50 @@ def test_ffill_bfill(self):
...
+        assert_panel_equal(a.bfill(axis=0), b0)
+
+        # method='bfill'
+        # axis=1
+        assert_panel_equal(a.bfill(axis=1), b1)
+
+        # method='bfill'
+        # axis=2
+        assert_panel_equal(a.bfill(axis=2), b2)
+
```
Comment (jreback) [inline pandas/tests/test_panel.py]: ideally also tests with a panel that has a non-numeric type as well (IOW, add a string to the above example); this should be a separate set of tests.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls comment if youd like to update / rebase and continue

**Pull request closed without merging**